packages/SwingSet/test/test-vpid-liveslots.js

```diff
@@ -300,52 +303,24 @@ async function doVatResolveCase23(t, which, mode) {
   // now it sends three() with the promise. For now, we expect the same vpid
   // as before.
 
-  // TODO: when we start to retire vpids, we expect to see vat-allocated
-  // 'p+7' (i.e. expectedP4) to be provided in args[], which will increment
-  // all the subsequent vat-allocated vpids:
-  //  * the resultSlot of three() will change to expectedP5
-  //  * the resultSlot of four() will change to expectedP6
-  //  * both subscriptions will change to match
-
   t.deepEqual(log.shift(), {
     type: 'send',
     targetSlot: target1,
     method: 'three',
-    args: capargs([slot0arg], [p1]), // this vpid will change
-    resultSlot: expectedP4,
+    args: capargs([slot0arg], [expectedP4]),
+    resultSlot: expectedP5,
   });
-  t.deepEqual(log.shift(), { type: 'subscribe', target: expectedP4 });
-
-  // TODO: when #823 lands, set this to false, and remove the dead code
-  const bug823broken = true;
+  t.deepEqual(log.shift(), { type: 'subscribe', target: expectedP5 });
 
-  if (bug823broken) {
-    // if #823 is still broken, we'll see a pipelined call to p1,
-    // unaware of how p1 was resolved
-    t.deepEqual(log.shift(), {
-      type: 'send',
-      targetSlot: p1,
-      method: 'four',
-      args: capargs([], []),
-      resultSlot: expectedP5,
-    });
-    t.deepEqual(log.shift(), { type: 'subscribe', target: expectedP5 });
-  } else {
-    // now, if we resolved p1 to a presence, the vat can send four() to it
-    // eslint-disable-next-line no-lonely-if
-    if (mode === 'presence') {
-      t.deepEqual(log.shift(), {
-        type: 'send',
-        targetSlot: target2,
-        method: 'four',
-        args: capargs([], []),
-        resultSlot: expectedP5,
-      });
-      t.deepEqual(log.shift(), { type: 'subscribe', target: expectedP5 });
-    }
-    // but if we resolved p1 to data or rejected it, four() is not sent
-    // (the vat creates an error for it locally)
-  }
+  t.deepEqual(log.shift(), {
+    type: 'send',
+    targetSlot: p1,
```

Comment:
yeah, if #823 were really fixed, then this send should not be going to the promise, because by the time `E(p1).four()` is called, `p1` should have been resolved to the `target2` presence. In fact `p1` should never be mentioned again after we get the `resolutionOf(p1)` earlier: that will fail because the kernel will have removed p1 by that point. As you noted in #823, there's more going on here than that one HandledPromise issue.

This might prevent landing this change, which would be a bummer.

Reply:
I think you are right that we shouldn't be seeing the message mention `p1`.  This works OK in the test because it's not actually resorting to the c-list to figure out what `p1` actually is, but I believe it would result in a failure if the vat were interacting with a real kernel.

Reply:
yeah. I did some more testing, and found out the behavior depends upon how many turns we stall, after resolving the promise but before we send `three()` and `four()`:

* 0 or 1 turns: everything gets pipelined to the original `p1` vpid, then the `syscall.resolve` happens at the very end
* 2 turns: `syscall.resolve` happens before three/four, so `three()` gets a new vpid, but `four` is sent to the old vpid, which will have been retired by the kernel by that point
* 3 turns: our local `p1` is resolved enough for liveslots to notice by the time `four()` is sent, so we see a `syscall.send` to the target that p1 was resolved to. (I changed the test to have it resolve to an imported object, so I could observe the four() syscall, since before it was resolving to something internal)

That 2-turns case is a problem, it will violate the contract. I see two tasks:

* figure out how to sanely run the test with varying numbers of stalls
* figure out the liveslots/HandledPromise behavior that is remembering the old vpid for too long

I'm thinking the HP needs to remember something other than the vpid, something that we look up at the time we do the `syscall.send` so that it will honor the same rules as all other sends. Maybe we should look the HP itself up in a table just before we do the send, to see what target it's associated with (allocating a new vpid if necessary?).

Reply:
I think we need to get to the bottom of why different numbers of stalls result in different behavior.  Keeping a table seems like a band-aid to get around our lack of understanding, and I fear if we don't understand what's going on this may come around to bite us later in some different context.

Reply:
this number-of-stalls stuff got resolved by changes in HandledPromise